src/pharmpy/plugins/nonmem/advan.py

```diff
@@ -548,31 +555,19 @@ def _advan12_trans(trans: str):
         )
 
 
-def _dosing(model: Model, control_stream: NMTranControlStream, dose_comp: int):
-    # FIXME Use lambda: model.dataset instead
-    def dataset():
-        if not hasattr(model, '_data_frame'):
-            return model._read_dataset(control_stream, raw=False)
-        return model._data_frame
-
-    return dosing(model.datainfo, dataset, dose_comp)
-
-
 def dosing(di: DataInfo, dataset: Callable[[], pd.DataFrame], dose_comp: int):
-    colnames = di.names
-
-    if 'RATE' in colnames and not di['RATE'].drop:
-        df = dataset()
-        if (df['RATE'] == 0).all():
-            return Bolus(sympy.Symbol('AMT'))
-        elif (df['RATE'] == -1).any():
-            return Infusion(sympy.Symbol('AMT'), rate=sympy.Symbol(f'R{dose_comp}'))
-        elif (df['RATE'] == -2).any():
-            return Infusion(sympy.Symbol('AMT'), duration=sympy.Symbol(f'D{dose_comp}'))
-        else:
-            return Infusion(sympy.Symbol('AMT'), rate=sympy.Symbol('RATE'))
-    else:
+    if 'RATE' not in di.names or di['RATE'].drop:
+        return Bolus(sympy.Symbol('AMT'))
+
+    df = dataset()
```

Comment:
Is the dataset callable to do lazy parsing?

Reply:
Exactly.

Reply:
Ok